### PR TITLE
feat: adding multichain 7702 auth support

### PR DESCRIPTION
**Pull request by vr16x**
## PR-Codex overview
This PR focuses on enhancing the `MEE` client functionality, particularly around authorization handling for multichain transactions and improving the testing of these features.

### Detailed summary
- Updated `getQuote` to accept an array of `authorizations` instead of a single `authorization`.
- Introduced helper functions `getDefaultMEENetworkUrl` and `getDefaultMEENetworkApiKey`.
- Modified `DEFAULT_PATHFINDER_URL` and `DEFAULT_PATHFINDER_API_KEY` to use the new helper functions.
- Enhanced `DelegationParams` to include `chainId` as a required field.
- Improved handling of multichain authorizations in `getQuote`, including validation for unique and same nonces.
- Added tests to verify automatic and manual injection of `7702` authorization for multichain transactions.
- Enhanced error handling for invalid authorizations in various scenarios.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/sdk/account/decorators/waitForTransactionReceipts.ts [hunk 1]
```diff
@@ -39,7 +39,10 @@ export const waitForTransactionReceipts = async (
 ): Promise<WaitForTransactionReceiptPayload> => {
   const receipts = await Promise.all(
     parameters.account.deployments.map(({ publicClient }, i) =>
-      publicClient.waitForTransactionReceipt({ hash: parameters.hashes[i] })
+      publicClient.waitForTransactionReceipt({
+        hash: parameters.hashes[i],
+        confirmations: 5
```
Comment (fichiokaku) [inline src/sdk/account/decorators/waitForTransactionReceipts.ts:44]: why 5?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 2]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+
+    // Auth is only being signed manually here for comparision purposes
+    const optimismSepoliaAuth =
+      await optimismSepoliaWalletClient.signAuthorization({
+        contractAddress: optimismSepoliaMcNexusVersion.implementationAddress
+      })
+
+    const meeClient = await createMeeClient({
+      account: mcNexus,
+      apiKey: "mee_3Zmc7H6Pbd5wUfUGu27aGzdf"
```
Comment (fichiokaku) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1311]: why not use default api key here?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 3]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+        amount: 1n,
+        tokenAddress: testnetMcTestUSDCP.addressOn(optimismSepolia.id)
+      }
+    })
+
+    const quote = await meeClient.getQuote({
+      sponsorship: true,
+      sponsorshipOptions: {
+        url: getDefaultMEENetworkUrl(true),
+        gasTank: {
```
Comment (fichiokaku) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1338]: can you please extract these 3 values to `getTestnetGasTank()` 

Will be easier for our clients to use when testing on staging

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 4]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+    })
+
+    const optimismSepoliaAuth =
+      await optimismSepoliaWalletClient.signAuthorization({
+        contractAddress: optimismSepoliaMcNexusVersion.implementationAddress
+      })
+
+    const meeClient = await createMeeClient({
+      account: mcNexus,
+      apiKey: "mee_3Zmc7H6Pbd5wUfUGu27aGzdf"
```
Comment (fichiokaku) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1549]: can we use default here?

**Review by fichiokaku — APPROVED** [review]

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 2 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1311]: Default API key is not sponsorship enabled. If we enable that ? we can remove this

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 4 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1549]: It is being used because of sponsorship. But will remove in next release with default key as sponsorship enabled one

File: src/sdk/account/decorators/waitForTransactionReceipts.ts (hunk 1 above)
Comment (vr16x) [reply src/sdk/account/decorators/waitForTransactionReceipts.ts:44]: 5 seems to be a best value to ensure the it is fast + good enough wait time to get the actual undelegation.

Earlier this is 1 and I don't want to go with less confirmations for this.

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 5]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+        }
+      },
+      delegate: true,
+      authorizations: [], // No auth is added. So the SDK should add auth message in all userOps with multichain context
+      instructions: [...baseSepoliaTransfer, ...optimismSepoliaTransfer]
+    })
+
+    expect(quote).toBeDefined()
+
+    expect(quote.userOps[0].userOp.initCode).to.eq("0x")
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1351]: is this a payment userOp? how is it processed if `initcode` is 0x and no authorizations?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 6]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+    expect(quote.userOps[2].eip7702Auth?.chainId).to.eq(optimismSepolia.id)
+    expect(quote.userOps[2].eip7702Auth?.nonce).to.eq(optimismSepoliaAuth.nonce)
+    expect(quote.userOps[2].eip7702Auth?.r).to.eq(optimismSepoliaAuth.r)
+    expect(quote.userOps[2].eip7702Auth?.s).to.eq(optimismSepoliaAuth.s)
+    expect(quote.userOps[2].eip7702Auth?.yParity).to.eq(
+      optimismSepoliaAuth.yParity
+    )
+  })
+
+  test("Should SDK ignore manually signed 7702 auth if the userOps doesn't exist for the chain", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1499]: I don't quite get the purpose of this test case. what does it test? the fact that the SDK may effectively reject an excess authorization and still return a valid quote?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 7]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+      version.implementationAddress
+    )
+    expect(quote.userOps[2].eip7702Auth?.chainId).to.eq(auth.chainId)
+    expect(quote.userOps[2].eip7702Auth?.nonce).to.eq(auth.nonce)
+    expect(quote.userOps[2].eip7702Auth?.r).to.eq(auth.r)
+    expect(quote.userOps[2].eip7702Auth?.s).to.eq(auth.s)
+    expect(quote.userOps[2].eip7702Auth?.yParity).to.eq(auth.yParity)
+  })
+
+  test("Should SDK use manually injected 7702 auth with chain id zero for all chains if the nonce are same", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1699]: I think test name should be rephrased, because it is no straightforward what is the difference with the previous one.

From expects I can see, that it expects auth to be added to paymentUserOp instead of user's userOp
everything else is the same as above
maybe this can be reflected in the name? or am I missing smthng?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 8]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+        multichain7702Auth: true,
+        authorizations: [baseSepoliaAuth, optimismSepoliaAuth],
+        instructions: [...baseSepoliaTransfer, ...optimismSepoliaTransfer],
+        feeToken: {
+          address: testnetMcTestUSDCP.addressOn(optimismSepolia.id),
+          chainId: optimismSepolia.id
+        }
+      })
+    ).rejects.toThrow(
+      "Invalid authorizations: The nonce for all the chains are zero and only one multichain authorization is expected"
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:2015]: what is the motivation to throw here?
auths are already signed at this point. 
Is the purpose to inform dev about suboptimal way of integration when dev notices this error at the testing time?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 9]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+        multichain7702Auth: true,
+        authorizations: [baseSepoliaAuth],
+        instructions: [...baseSepoliaTransfer],
+        feeToken: {
+          address: testnetMcTestUSDCP.addressOn(baseSepolia.id),
+          chainId: baseSepolia.id
+        }
+      })
+    ).rejects.toThrow(
+      "Invalid authorizations: Multichain authorization should be signed with chain ID zero"
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:2072]: I do not get it. the superTx is for baseSepolia. why baseSepolia auth doesn;t work here?

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 10]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+          address: testnetMcTestUSDCP.addressOn(optimismSepolia.id),
+          chainId: optimismSepolia.id
+        }
+      })
+    ).rejects.toThrow(
+      "Authorizations are missing for the following chains: 11155420"
+    )
+  })
+
+  test("Should SDK throw an error for single chain authorization if the multichain auth is passed", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:2137]: don't get the naming.  

> single chain authorization if the multichain auth

single chain or multichain?

and why it reverts? auth is multichain with chainid 0

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 11]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+          address: testnetMcTestUSDCP.addressOn(optimismSepolia.id),
+          chainId: optimismSepolia.id
+        }
+      })
+    ).rejects.toThrow(
+      "Authorizations are missing for the following chains: 11155420, 84532"
+    )
+  })
+
+  test("Should execute multiple 7702 delegation supertx with manual auth and single chain auth with sponsorship", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts]:
```suggestion
  test("Should execute multiple 7702 delegation supertx with manual single chain authorisations with sponsorship", async () => {
```

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 12]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+    if (isDelegated) {
+      await mcNexus.unDelegate()
+
+      isDelegated = await mcNexus.isDelegated()
+    }
+
+    expect(isDelegated).toBe(false)
+  })
+
+  test("Should execute multiple 7702 delegation supertx with authomatic auth and single chain auth", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:2323]:
```suggestion
  test("Should execute multiple 7702 delegation supertx with automatic single chain auths", async () => {
```

File: src/sdk/clients/decorators/mee/getQuote.test.ts [hunk 13]
```diff
@@ -1254,4 +1255,1341 @@ describe("mee.getQuote", () => {
...
+    if (isDelegated) {
+      await mcNexus.unDelegate()
+
+      isDelegated = await mcNexus.isDelegated()
+    }
+
+    expect(isDelegated).toBe(false)
+  })
+
+  test("Should execute multichain 7702 delegation supertx with manual auth", async () => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts]:
```suggestion
  test("Should execute multichain 7702 delegation supertx with manual multichain auth", async () => {
```

File: src/sdk/clients/decorators/mee/getQuote.test.ts
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.test.ts:1]: The file is huge and IMO could benefit from improving the reused code ratio. I believe a lot of tests just use very same or similar code snippets which are quite long

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 5 above)
Comment (filmakarov) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1351]: ah, I see, this is because of sponsorship

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 7 above)
Comment (filmakarov) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1699]: I mean, the difference is sponsorship, prolly it should be reflected in the naming

File: src/sdk/clients/decorators/mee/getQuote.ts [hunk 14]
```diff
@@ -555,21 +568,193 @@ export const getQuote = async (
...
+          // If noncesAndChainIdsWithUniqueNonces length is zero ? It means all the nonces are same and can be used for multichain
+          // It is expected to pass only one auth from outside the SDK.
+          if (
+            noncesAndChainIdsWithUniqueNonces.length === 0 &&
+            authorizations.length > 1
+          ) {
+            throw new Error(
+              "Invalid authorizations: The nonce for all the chains are zero and only one multichain authorization is expected"
+            )
+          }
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.ts:637]: is this going to be fixed before we merge? since it can cause false negatives as Filip mentioned

File: src/sdk/clients/decorators/mee/getQuote.ts [hunk 15]
```diff
@@ -555,21 +568,193 @@ export const getQuote = async (
...
+
+          // If multichain nonce are not same and custom auth are passed ? The auth should be sufficient orelse error will be thrown
+          if (noncesAndChainIdsWithUniqueNonces.length > 0) {
+            const missingAuthsByChainId: number[] = []
+
+            for (const { chainId } of noncesAndChainIdsWithUniqueNonces) {
+              const isAuthExist = authorizations.some((auth) => {
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.ts]:
```suggestion
              const isAuthProvided = authorizations.some((auth) => {
```

File: src/sdk/clients/decorators/mee/getQuote.ts [hunk 16]
```diff
@@ -555,21 +568,193 @@ export const getQuote = async (
...
+              )
+              throw new Error(
+                `Invalid authorizations: The nonce for some of the chains are same. Missing multichain authorization for the following chains: ${chainIds.join(", ")}`
+              )
+            }
+          }
+        }
+
+        for (const chainId of sprtxChainIds) {
+          const [isMultichainAuth] = noncesAndChainIdsWithSameNonces.filter(
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.ts:676]:
```suggestion
          const [isMultichainAuthRequired] = noncesAndChainIdsWithSameNonces.filter(
```

File: src/sdk/clients/decorators/mee/getQuote.ts [hunk 17]
```diff
@@ -612,20 +797,54 @@ export const getQuote = async (
...
+            // Apply existing multichain auth where the chain Ids were same. So no need multiple auths to be signed
+            if (multichainEIP7702Auth) {
+              initDataOrUndefined = {
+                eip7702Auth: multichainEIP7702Auth
+              }
+            } else {
+              // This multichain auth will be used for the current chains and other chains which has the same nonce
+              multichainEIP7702Auth = await prepare7702Auth(
+                nexusAccount,
+                Number(chainId),
```
Comment (filmakarov) [inline src/sdk/clients/decorators/mee/getQuote.ts:827]: chainId expected to be 0 here?

Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.ts:827]: This is a common helper. It decides inside the util to use zero vs this chainId

File: src/sdk/clients/decorators/mee/getQuote.ts (hunk 14 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.ts:637]: Discussed in threads

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 10 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:2137]: If you don't enable multichain7702Auth = true - The SDK always process for single chain auth. In this case, if you pass multichain auth with chainId zero ? it throws an error.

It encourage people to use individual chainId if multichain auth is not opted.

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 6 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1499]: Yeah correct.

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 7 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1699]: So I have a convention where sponsorship is a special case. If the tx has a sponsorship, it will be included in test title orelse it is always non sponsored

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 8 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:2015]: Yeah correct, this is the strict validation to correct the integration.

If we don't do this validation, SDK will be open for anything to be passed and it is very hard to handle all the possible edge cases. So we should have some rules in place or use non multichain version

File: src/sdk/clients/decorators/mee/getQuote.test.ts (hunk 9 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:2072]: multichain7702Auth: true

SDK expects the auth to be signed with zero chainId. If not ? It will throw an error

File: src/sdk/clients/decorators/mee/getQuote.test.ts
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.test.ts:1]: Will refactor in next iteration if that's okay

File: src/sdk/clients/decorators/mee/getQuote.ts (hunk 16 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.ts:676]: This is a object and not a boolean. So I am checking whether this is a multiChainAuth or not and required keyword is not mandatory. I think not a big deal

File: src/sdk/clients/decorators/mee/getQuote.ts (hunk 15 above)
Comment (vr16x) [reply src/sdk/clients/decorators/mee/getQuote.ts]: Fixed
